velox/exec/Window.cpp

```diff
@@ -374,19 +384,55 @@ void Window::callApplyForPartitionRows(
     // as WindowFunction only sees one partition at a time.
     rawPeerStarts[j] = peerStartRow_ - firstPartitionRow;
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
+  }
 
-    for (auto w = 0; w < numFuncs; w++) {
-      auto frameEndPoints = findFrameEndPoints(
-          w,
-          firstPartitionRow,
-          lastPartitionRow,
-          peerStartRow_,
-          peerEndRow_ - 1,
-          i);
-      rawFrameStartBuffers[w][j] = frameEndPoints.first - firstPartitionRow;
-      rawFrameEndBuffers[w][j] = frameEndPoints.second - firstPartitionRow;
+  auto updateFrameBounds = [&](vector_size_t* frameBound,
+                               core::WindowNode::BoundType boundType,
+                               core::WindowNode::WindowType type,
+                               bool isStartBound) -> void {
+    switch (boundType) {
+      case core::WindowNode::BoundType::kUnboundedPreceding:
+        findRowUnboundedPreceding(frameBound, numRows);
+        break;
+      case core::WindowNode::BoundType::kCurrentRow: {
+        if (type == core::WindowNode::WindowType::kRange) {
+          if (isStartBound) {
+            findCurrentRowFrameStart(
+                frameBound, firstPartitionRow, rawPeerStarts, numRows);
+          } else {
+            findCurrentRowFrameEnd(
+                frameBound, firstPartitionRow, rawPeerEnds, numRows);
+          }
+        } else {
+          VELOX_NYI("Not supported");
+        }
+        break;
+      }
+      case core::WindowNode::BoundType::kPreceding:
+      case core::WindowNode::BoundType::kFollowing:
+        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kUnboundedFollowing:
```

Comment:
You can move this case just below kUnboundedPreceding as they are very similar.

Reply:
Sure, thanks for the suggestion!